cli/disk/manager.go

```diff
@@ -0,0 +1,309 @@
+// © Broadcom. All Rights Reserved.
+// The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package disk
+
+import (
+	"context"
+	"errors"
+	"os"
+	"time"
+
+	"github.com/vmware/govmomi/cli/flags"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/govmomi/vslm"
+	vslmtypes "github.com/vmware/govmomi/vslm/types"
+)
+
+// Manager provides a layer for switching between Virtual Storage Object Manager (VSOM)
+// and Virtual Storage Lifecycle Manager (VSLM). The majority of VSOM methods require a
+// Datastore param, and most VSLM methods do not as it uses the "Global Catalog".
+// VSOM was introduced in vSphere 6.5 (11/2016) and VSLM in vSphere 6.7 U2 (04/2019).
+// The govc disk commands were introduced prior to 6.7 U2 and continue to use VSOM
+// when an optional Datastore (-ds flag) is provided. Otherwise, VSLM global methods
+// are used when a Datastore is not specified.
+// Also note that VSOM methods can be used when connected directly to an ESX hosts,
+// but VSLM global methods require vCenter.
+// A disk managed by these methods are also known as a "First Class Disk" (FCD).
+type Manager struct {
+	Client              *vim25.Client
+	Datastore           *object.Datastore
+	ObjectManager       *vslm.ObjectManager
+	GlobalObjectManager *vslm.GlobalObjectManager
+}
+
+var (
+	// vslm does not have a PropertyCollector, the Wait func polls via VslmQueryInfo
+	taskTimeout = time.Hour
+
+	// Some methods require a Datastore param regardless using vsom or vslm.
+	// By default we use Global vslm for those methods, use vsom when this is "false".
+	useGlobal = os.Getenv("GOVC_GLOBAL_VSLM") != "false"
+)
+
+func NewManagerFromFlag(ctx context.Context, cmd *flags.DatastoreFlag) (*Manager, error) {
+	c, err := cmd.Client()
+	if err != nil {
+		return nil, err
+	}
+
+	ds, err := cmd.DatastoreIfSpecified()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewManager(ctx, c, ds)
+}
+
+func NewManager(ctx context.Context, c *vim25.Client, ds *object.Datastore) (*Manager, error) {
+	m := &Manager{
+		Client:        c,
+		Datastore:     ds,
+		ObjectManager: vslm.NewObjectManager(c),
+	}
+
+	if ds == nil {
+		if err := m.initGlobalManager(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	return m, nil
+}
+
+func (m *Manager) initGlobalManager(ctx context.Context) error {
+	if m.GlobalObjectManager == nil {
```

Comment:
Should this use `sync.Once` anywhere?

Reply:
Might just get rid of this deferred initGlobalManager in the follow up to add metadata and query support. Instead will always create `vslm.NewClient` in `NewManager`, unless we're connected directly to ESX.
Also, only consumer of this disk.Manager is govc. I think most applications should directly just use vslm.Client against vCenter. Maybe we do the same in govc at some point and this Manager goes away completely.